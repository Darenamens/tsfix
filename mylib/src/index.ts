import path from "path";
import { CodeFixAction, Diagnostic, FileTextChanges, getDefaultFormatCodeSettings, TextChange } from "typescript";
import { createProject, Project } from "@ts-morph/bootstrap";
import os from "os";
var _ = require('loadash/array')

const tsConfigFilePath = path.resolve(__dirname, "../test/exampleTest/tsconfig.json");

export async function codefixProject() {
  const firstPass = await applyCodefixesOverProject();

  // if overlap/non executed changes for some reason, redo process 
  if (!firstPass) {
    // maybe in applycodefixesoverproject we have an exit if no changes?
    // maybe emit some sort of diagnostic/statement? 
    // might need some other type/flag with more options besides boolean 
    applyCodefixesOverProject();
  }
  
  return; 
}



export async function applyCodefixesOverProject() : Promise<readonly boolean> {
  // get project object
  let project = await getProject();
  
  // pull all codefixes
  const diagnosticsPerFile = await getDiagnostics(project);

  // format codefixes -- pull text edits
  // since the type of change is associated with the codefix or the diagnostic,
  //   we should filter for which codefixes we want in this step, probably write 
  //   the filter into getCodeFixesForFile()
  const codefixesPerFile = diagnosticsPerFile.map(function (d) {
    return (getCodeFixesForFile(project, d));
 });
 const codefixes = _.flatten(codefixesPerFile);

  // format codefixes -- put into dictionary by file
  //                  -- file is specified by fileChange so this sorting is done
  //                        within the getTextChangesForCodeFix 
 const textChangesPerFile = getTextChangesForCodeFix( codefixes );
  //  .map(function (fixList) {
  //    return getTextChangesForCodeFix(fixList);
  //  });


  // edit each file, filter for overlap


  for (d in diagnosticsPerFile){
    // applyCodefixes(project, project.getFileNames)

  }
}

function getFileFromDiagnostic(d : Diagnostic) : string {
  return d.source;
}

async function getProject() : Promise<Project> {
  return createProject({ tsConfigFilePath });
}

export function getFileNames(project: Project): string[] {
  const files = project.getSourceFiles();
  const fileNames = files.map(function(files) {
    return files.fileName;
  })
  // console.log(fileNames);
  return fileNames;
}

export function getDiagnostics(project: Project): (readonly Diagnostic[])[] {
  const diagnostics = project.getSourceFiles().map(function(file) {
    return project.createProgram().getSemanticDiagnostics(file);
  });
  return diagnostics;
}

function getCodeFixesForFile(project: Project, diagnostics : readonly Diagnostic[]) : readonly CodeFixAction[] {
  const service = (project).getLanguageService();
  const codefixes = (<CodeFixAction[]>[]).concat.apply([], diagnostics.map(function (d) {
    if (d.file && d.start !== undefined && d.length !== undefined) {
      return service.getCodeFixesAtPosition(
        d.file.fileName, 
        d.start, 
        d.start + d.length, 
        [d.code],
        getDefaultFormatCodeSettings(os.EOL),
        {});
    } else {
      return [];
    } 
  })).filter(d => d !== undefined);
  return codefixes;
}

function getFileTextChangesFromCodeFix(codefix : CodeFixAction) : readonly FileTextChanges[] {
  return codefix.changes;
}

function getTextChangesForFileFromFileChange(change : FileTextChanges) : readonly TextChange[] {
  // do we need to make sure all codefixes apply to same file? my thinking is yes.
  return change.textChanges;
}

function getTextChangesForCodeFix(codefixes: readonly CodeFixAction[]) : readonly TextChange[] {
  // get 1d list of fileChanges list from list of codefixactions
  const changesToFile = (<FileTextChanges[]>[]).concat.apply([], codefixes.map(function (fix) {
    return getFileTextChangesFromCodeFix(fix);
  }) );
  // const textChanges = (<TextChange[]>[]).concat.apply([], changesToFile.map(function (fix) {
  //   return getTextChangesForFileFromFileChange(fix);
  // }) );

  let textChanges = new Map();

  return textChanges;
}

function collectTextChanges(textChanges: readonly TextChange[]) {

}

function getFileOfTextChange(change: TextChange) {

}

function applyCodefixes(project: Project, filePath: string, diagnostics : Diagnostic[]) {
  const codefixes = getCodeFixesForFile(project, diagnostics);
  // get list of text changes
  const textChanges = getTextChangesForCodeFix(codefixes);
  const originalFile = project.getSourceFile(filePath); 
  if (originalFile === undefined){
    // if file cannot be found, return undone text changes ->  throw error?
    return textChanges;
  }
  
  let originalFileText = originalFile.text;

  // do the text changes
  // function should return changes that aren't done, and new file contents
  const newFileString = doTextChanges(originalFileText, textChanges);
  // return string of new file contents + all changes that weren't done?
  return newFileString;
}

async function doTextChanges(fileText : string, textChanges: readonly TextChange[]) {
  // sort text changes by TextChange, by now we have made sure that they all apply to the correct file

  // iterate through codefixes from back
  for (let i = textChanges.length - 1; i >= 0; i--) {
    // apply each codefix
    fileText = await doTextChange(fileText, textChanges[i]);
  }

  return fileText;
}

async function doTextChange(currentFileText : string, change : TextChange) {
  const prefix = currentFileText.substring(0, change.span.start);
  const middle = change.newText;
  const suffix = currentFileText.substring(change.span.start + change.span.length);
  return prefix + middle + suffix;
}




