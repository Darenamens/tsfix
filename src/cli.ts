#!/usr/bin/env node

import { writeFileSync } from 'fs';
import yargs, { boolean, string } from 'yargs';
import path from "path";
import { Options, codefixProject, Host } from '.';

const argv = yargs(process.argv.slice(2))
            .scriptName("codefix")
            .usage("$0 -t tsconfig.json -o /output -e err#")
            .option("t", {
                alias: "tsconfig",
                description: "name or path to project's tsconfig",
                type: "string",
                nargs: 1
            })
            .option("file", {
                description: "files to codefix. Not implemented yet.",
                type: "string"
            })
            .option("e", {
                alias: "errorCode",
                describe: "the error code (number)",
                type: "number"
            })
            .option("f", {
                alias: "fixName",
                describe: "names of codefixes to apply",
                type: "string"
            }) 
            .option("r", {
               alias: "replace",
               describe: "Default: True. Set flag to false if code-fixed code should be emitted into original file (overwrites original file)",
               type: "boolean"
           })
           .option("o", {
            alias: "outputDirectory", 
            describe: "the output directory. Only used if -r is explicitly false (--no-r)",
            type: "string"
            }).argv;

export function makeOptions(cwd:string, argv:any) : Options { // Tested
    const {t, e, o, f, r} = argv;
    let tsconfigPath = (t===undefined) ? path.resolve(cwd, "tsconfig.json") : path.resolve(t);
    return  {
        tsconfigPath : tsconfigPath,
        replace :  (r===true||r===undefined),
        outputFolder : (r===true||r===undefined) ? path.dirname(tsconfigPath) : (o===undefined) ? path.resolve(path.dirname(tsconfigPath), "../Output") : path.resolve(o),
        errorCode : (e===undefined) ? [] : (typeof e === "number") ? [e] : e,
        fixName : (f===undefined) ? [] :  (typeof f === "string") ? [f] : f,
    }
}

const cliHost: Host = {
    log: console.log.bind(console),
    writeFile: (fileName, content) => writeFileSync(fileName, content, 'utf8'),
};



// runCli('--tsconfig ./blah/whatever/tsconfig.json --fixNames 23423', testHost);
// logs: [ ... ]
// changedFiles: [ ... ]
const opt = makeOptions(process.cwd(), argv);
console.log(opt.tsconfigPath);
console.log(opt.outputFolder);
codefixProject(opt, cliHost);
//DONE: print out how many errors matched
//fixes matched 

// print out files changed

// overlapping errors (matched changes that were not applied)

