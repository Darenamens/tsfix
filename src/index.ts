import path from "path";
import { CodeFixAction, Diagnostic, FileTextChanges, getDefaultFormatCodeSettings, getOriginalNode, TextChange } from "typescript";
import { createProject, Project } from "@ts-morph/bootstrap";
import os from "os";
import fs from "fs";
import _ from "lodash";
// export const tsConfigFilePathDefault = path.resolve(__dirname, "../test/exampleTest/tsconfig.json");
// const outputFolderDefault = path.resolve(__dirname, "../test/exampleTestOutput");

export interface Options {
  tsconfigPath: string;
  outputFolder: string;
  errorCode: number[];
  fixName: string[];
}



export async function codefixProject(opt:Options) {
  const firstPass = await applyCodefixesOverProject(opt);

  // if overlap/non executed changes for some reason, redo process 
  if (!firstPass) {
    // maybe in applycodefixesoverproject we have an exit if no changes?
    // maybe emit some sort of diagnostic/statement? 
    // might need some other type/flag with more options besides boolean 
    return applyCodefixesOverProject(opt);
  }

  return firstPass;
}



export async function applyCodefixesOverProject(opt: Options): Promise<boolean> {
  // tsconfigPath: string, errorCode?: number|undefined
  // get project object
  let project = await getProject(opt.tsconfigPath);

  // pull all codefixes.
  const diagnosticsPerFile = await getDiagnostics(project);

  // pull codefixes from diagnostics.  If errorCode is specified, only pull fixes for that/those errors. 
  //    Otherwise, pull all fixes
  const codefixesPerFile = diagnosticsPerFile.map(function (d) {
    return (getCodeFixesForFile(project, d, opt)); 
  });
  const codefixes = <CodeFixAction[]>_.flatten(codefixesPerFile);

  // filter for codefixes if applicable, then 
  //    organize textChanges by what file they alter
  const textChangesByFile = getTextChangeDict(codefixes, opt);

  // edit each file
  let leftoverChanges = doAllTextChanges(project, textChangesByFile, opt);
  // figure out returns alater....
  return leftoverChanges;
}

export async function getProject(tsConfigFilePath: string): Promise<Project> {
  return createProject({ tsConfigFilePath });
}

export function getDiagnostics(project: Project): (readonly Diagnostic[])[] {
  const diagnostics = project.getSourceFiles().map(function (file) {
    return project.createProgram().getSemanticDiagnostics(file);
  });
  return diagnostics;
}

export function getCodeFixesForFile(project: Project, diagnostics: readonly Diagnostic[], opt: Options): readonly CodeFixAction[] {
  const service = (project).getLanguageService();
  const filteredDiagnostics = filterDiagnosticsByErrorCode(diagnostics,opt);
  const codefixes = (<CodeFixAction[]>[]).concat.apply([], filteredDiagnostics.map(function (d) {
    if (d.file && d.start !== undefined && d.length !== undefined) {
      return service.getCodeFixesAtPosition(
        d.file.fileName,
        d.start,
        d.start + d.length,
        [d.code],
        getDefaultFormatCodeSettings(os.EOL),
        {});
    } else {
      return [];
    }
  })).filter(d => d !== undefined);
  return codefixes;
}

export function filterDiagnosticsByErrorCode(diagnostics: readonly Diagnostic[], opt:Options): readonly Diagnostic[] {
  // if errorCodes were passed in, only use the specified errors
  if (opt.errorCode.length !== 0) {
    return _.filter(diagnostics, function (d) {return opt.errorCode.includes(d.code)});
    // TODO: maybe add more sophisiticated method of error matching?
  }
  // otherwise, use all errors
  return diagnostics;
}

function getFileTextChangesFromCodeFix(codefix: CodeFixAction): readonly FileTextChanges[] {
  return codefix.changes;
}

// function getTextChangesFromFileChanges(change : FileTextChanges) : readonly TextChange[] {
//   // do we need to make sure all codefixes apply to same file? my thinking is yes.
//   return change.textChanges;
// }

// function getTextChangesForCodeFix(codefix: CodeFixAction) : readonly TextChange[] {
//   // get 1d list of fileChanges list from list of codefixactions
//   const changesFromCodefix = codefix.changes;
//   const textChanges = _.flatten(changesFromCodefix.map(getTextChangesFromFileChanges));
//   return textChanges;
// }

export function getTextChangeDict(codefixes: readonly CodeFixAction[], opt: Options): Map<string, TextChange[]> {
  let textChangeDict = new Map<string, TextChange[]>();

  const filteredFixes = filterCodeFixesByFixName(codefixes, opt);
  for (let i = 0; i < filteredFixes.length; i++) {
    const fix = filteredFixes[i];
    const changes = getFileTextChangesFromCodeFix(fix);

    for (let j = 0; j < changes.length; j++) {
      let change = changes[j];
      let [key, value] = getFileNameAndTextChangesFromCodeFix(change);

      const prevVal = textChangeDict.get(key);
      if (prevVal === undefined) {
        textChangeDict.set(key, value);
      } else {
        textChangeDict.set(key, prevVal.concat(value));
      }
    }
  }

  return textChangeDict;
}

function filterCodeFixesByFixName(codefixes: readonly CodeFixAction[], opt: Options): CodeFixAction[] {
  return codefixes.filter(function (codefix) {return  opt.fixName.includes(codefix.fixName);});
}

function getFileNameAndTextChangesFromCodeFix(ftchanges: FileTextChanges): [string, TextChange[]] {
  return [ftchanges.fileName, [...ftchanges.textChanges]];
}

function doAllTextChanges(project: Project, textChanges: Map<string, TextChange[]>, opt: Options): boolean {
  let leftoverChanges = false;
  textChanges.forEach((fileFixes, fileName) => {
    const sourceFile = project.getSourceFile(fileName);

    if (sourceFile !== undefined) {
      const originalFileContents = sourceFile.text;
      // let fileFixes = textChanges.get(fileName);

      // if (fileFixes === undefined) {
      //   fileFixes = <TextChange[]> [];
      // }

      // collision is true if there were changes that were not applied 
      // also performs the writing to the file
      let [collision, newFileContents] = applyCodefixesInFile(originalFileContents, fileFixes);
      writeToFile(fileName, newFileContents, opt);
      if (collision) {
        leftoverChanges = collision;
      }
    }

    else {
      throw new Error('file ' + fileName + ' not found in project');
    }
  });

  return leftoverChanges;
}

function applyCodefixesInFile(originalContents: string, textChanges: TextChange[]):  [ boolean, string] {
  // sort textchanges by start
  const sortedFixList = textChanges.sort((a, b) => a.span.start - b.span.start);

  // take some sort of issue (or none) with overlapping textchanges
  const [filteredFixList, notAppliedFixes] = filterOverlappingFixes(sortedFixList);

  // apply all remaining textchanges
  const newFileContents = applyChangestoFile(originalContents, filteredFixList);
  
  // return 
  // if all fixes have been applied, then it is False that we expect to do another pass
  return [notAppliedFixes.length === 0, newFileContents];
}


export function filterOverlappingFixes(sortedFixList: TextChange[]): [TextChange[], TextChange[]] { //TODO: 1 test this function
  let filteredList = <TextChange[]>[];
  let droppedList = <TextChange[]>[];
  let currentEnd = -1;
  // why is 'fix' in the line below a string[], while sortedFixList is Textchange[]?
  // for (const fix in sortedFixList) {
  //   filteredList + [fix];
  // }
  for (let i = 0; i < sortedFixList.length; i++) {
    let fix = sortedFixList[i];
    if (fix.span.start > currentEnd) {
      filteredList.push(fix);
      currentEnd = fix.span.start + fix.span.length;
    } else {
      droppedList.push(fix);
    }
  }
  return [filteredList, droppedList];
}

function applyChangestoFile(originalContents: string, fixList: TextChange[]): string {
  // maybe we want to have this and subsequent functions to return a diagnostic
  // function expects fixList to be already sorted and filtered
  const newFileContents = doTextChanges(originalContents, fixList);
  return newFileContents;
}

export function doTextChanges(fileText: string, textChanges: readonly TextChange[]): string {
  // does js/ts do references? Or is it always a copy when you pass into a function
  // iterate through codefixes from back
  for (let i = textChanges.length - 1; i >= 0; i--) {
    // apply each codefix
    fileText = doTextChangeOnString(fileText, textChanges[i]);
  }
  return fileText;
}

export function doTextChangeOnString(currentFileText: string, change: TextChange): string {
  const prefix = currentFileText.substring(0, change.span.start);
  const middle = change.newText;
  const suffix = currentFileText.substring(change.span.start + change.span.length);
  return prefix + middle + suffix;
}


function writeToFile(fileName: string, fileContents: string, opt: Options): void {
  const writeToFileName = getOutputFilePath(fileName, opt);
  const writeToDirectory =getDirectory(writeToFileName)
  if (!fs.existsSync(writeToDirectory)) {
    createDirectory(writeToDirectory);
  }
  fs.writeFileSync(writeToFileName , fileContents);
}

function createDirectory(directoryPath: string) {
  fs.mkdir(directoryPath, {recursive :true}, () => {});
}


function getOutputFilePath(filePath: string, opt: Options): string {
  const fileName = getRelativePath(filePath, opt);
  return path.resolve(opt.outputFolder, fileName);
}

export function getFileName(filePath: string): string {
  return filePath.replace(/^.*[\\\/]/, '');
}

export function getDirectory(filePath:string) :string {
  return filePath.substring(filePath.length - getFileName(filePath).length)
}

function getOutputBaseFolder( opt: Options):string {
  return opt.outputFolder;
}

export function getRelativePath(filePath: string, opt: Options): string{
  const baseFolder = opt.tsconfigPath.substring(opt.tsconfigPath.length - "tsconfig.json".length);
  return filePath.substring(baseFolder.length, filePath.length);
}
