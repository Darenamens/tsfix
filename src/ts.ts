import fs from "fs";
import path from "path";
import importCwd from "import-cwd";
import type { LanguageService, LanguageServiceHost, ParseConfigFileHost, Program } from "typescript";
import { parseInt } from "../../../../Users/t-isabelduan/TS-transform-project/node_modules/@types/lodash";

function isTypeScriptVersionSupported(major: number, minor: number) {
  if (major < 3) return false;
  if (major < 4) return minor >= 5;
  return true;
}

type TypeScript = typeof import("typescript");

export function loadTypeScript(): TypeScript {
  try {
    const ts = importCwd("typescript") as typeof import("typescript");
    const [major, minor] = ts.versionMajorMinor.split('.');
    if (isTypeScriptVersionSupported(parseInt(major, 10), parseInt(minor, 10))) {
      return ts;
    }
  } catch {}
  return require("typescript");
}

export interface CreateProjectOptions {
  tsConfigFilePath: string;
}

export interface Project {
  ts: TypeScript;
  languageService: LanguageService;
  program: Program;
}

export function createProject(options: CreateProjectOptions): Project | undefined {
  const ts = loadTypeScript();
  const parseConfigHost: ParseConfigFileHost = {
    fileExists: ts.sys.fileExists,
    getCurrentDirectory: ts.sys.getCurrentDirectory,
    readDirectory: ts.sys.readDirectory,
    readFile: ts.sys.readFile,
    useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
    onUnRecoverableConfigFileDiagnostic: diagnostic => {
      const message = ts.formatDiagnosticsWithColorAndContext([diagnostic], {
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: ts.sys.getCurrentDirectory,
        getNewLine: () => ts.sys.newLine
      });
      // TODO: let CLI choose how to handle error instead of throwing here
      throw new Error(message);
    }
  }

  const commandLine = ts.getParsedCommandLineOfConfigFile(path.resolve(options.tsConfigFilePath), undefined, parseConfigHost);
  if (!commandLine) return undefined;

  const languageServiceHost: LanguageServiceHost = {
    getCompilationSettings: () => commandLine.options,
    getProjectReferences: () => commandLine.projectReferences,
    getCurrentDirectory: ts.sys.getCurrentDirectory,
    getDefaultLibFileName: ts.getDefaultLibFileName,
    fileExists: ts.sys.fileExists,
    readFile: ts.sys.readFile,
    readDirectory: ts.sys.readDirectory,
    directoryExists: ts.sys.directoryExists,
    getDirectories: ts.sys.getDirectories,
    getScriptFileNames: () => commandLine.fileNames,
    getScriptVersion: () => "0", // Get a new project if files change
    getScriptSnapshot: fileName => {
      if (!fs.existsSync(fileName)) {
        return undefined;
      }

      return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString());
    },
  };

  const languageService = ts.createLanguageService(languageServiceHost);
  const program = languageService.getProgram();
  if (!program) return undefined;

  return { ts, languageService, program };
}

function parseInt(_major: any, _arg1: number): number {
    throw new Error("Function not implemented.");
}
