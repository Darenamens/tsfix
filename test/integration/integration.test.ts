import fs from "fs";
import path from "path";
import { codefixProject} from "../../src";
import { makeOptions } from "../../src/cli";
import { normalizeSlashes, TestHost } from "./testHost";
import {addSerializer, toMatchSpecificSnapshot} from "jest-specific-snapshot";

async function baselineCLI(cwd: string, args: string[]) {
  const host = new TestHost(cwd);
  const options = makeOptions(cwd, args);
  await codefixProject(options, host);
  
  const snapshot = {
    cwd: normalizeSlashes(path.relative(__dirname, cwd)),
    args,
    logs: host.getLogs(),
    filesWritten: host.getFilesWritten(),
  };
  
  return snapshot;
}

addSerializer({
  test(snapshot: { cwd: any; args: any; logs: any; filesWritten: any; }){
    return snapshot.cwd && snapshot.args && snapshot.logs && snapshot.filesWritten;
  },
  print(snapshot: { cwd: any; args: any; logs: any; filesWritten: any; }){
    function replacer(_, value:any) {
      if(value instanceof Map) {
        return {
          dataType: 'Map',
          value: Array.from(value.entries()), // or with spread: value: [...value]
        };
      } else {
        return value;
      }
    }
    return JSON.stringify(snapshot, replacer,2);
  }
})



const cases = fs.readdirSync(path.resolve(__dirname, "cases")).flatMap(dirName => {
  const commands = fs.readFileSync(path.resolve(__dirname, "cases", dirName, "cmd.txt"), "utf8").split(/\r?\n/);
  // Split cmd.txt by line, then into space-separated args, and leave off the leading `ts-fix`
  return commands.filter(c => c.trim().length > 0).map(c => c.split(" ").slice(1)).map((args): [string, string[]] => [dirName, args]);
});

describe("integration tests", () => {
  test.each(cases)("%s %#", async (dirName, args) => {
    const cwd = path.resolve(__dirname, "cases", dirName);
    const snapshot = await baselineCLI(path.posix.normalize(cwd), args);
   expect(snapshot).toMatchSpecificSnapshot(path.resolve(__dirname, '__snapshots__', dirName+".shot"));

  });
});

