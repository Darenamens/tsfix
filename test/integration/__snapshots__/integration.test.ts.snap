// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests addMissingMember 0 1`] = `
Object {
  "args": Array [
    "-e",
    "2339",
    "--write",
  ],
  "cwd": "cases\\\\addMissingMember",
  "filesWritten": Map {
    "types.ts" => "interface I {
  two: any;
  one: boolean;
}",
  },
  "logs": Array [
    "typescript version 4.3.5",
    "found 2 diagnostics with code 2339",
    "found 4 codefixes",
    "Updated types.ts",
    "Performing second passthrough",
    "found 2 diagnostics with code 2339",
    "found 4 codefixes",
    "Updated types.ts",
  ],
}
`;

exports[`integration tests addMissingOverride 1 1`] = `
Object {
  "args": Array [
    "-e",
    "4114",
    "--write",
  ],
  "cwd": "cases\\\\addMissingOverride",
  "filesWritten": Map {
    "index.ts" => "class Base {
  m() {}
}

class Derived extends Base {
  override m() {}
}

class MoreDerived extends Derived {
  override m() {}
}
",
  },
  "logs": Array [
    "typescript version 4.3.5",
    "found 2 diagnostics with code 4114",
    "found 2 codefixes",
    "Updated index.ts",
  ],
}
`;
